cmd/kubeadm/app/phases/addons/proxy/proxy.go

```diff
@@ -46,28 +48,136 @@ const (
 	KubeProxyConfigMapRoleName = "kube-proxy"
 )
 
+var PrintManifestFlag bool = false
+
 // EnsureProxyAddon creates the kube-proxy addons
-func EnsureProxyAddon(cfg *kubeadmapi.ClusterConfiguration, localEndpoint *kubeadmapi.APIEndpoint, client clientset.Interface) error {
-	if err := CreateServiceAccount(client); err != nil {
-		return errors.Wrap(err, "error when creating kube-proxy service account")
+func EnsureProxyAddon(cfg *kubeadmapi.ClusterConfiguration, localEndpoint *kubeadmapi.APIEndpoint, client clientset.Interface, out io.Writer) error {
+	if !PrintManifestFlag {
+		if err := CreateServiceAccount(client); err != nil {
+			return errors.Wrap(err, "error when creating kube-proxy service account")
+		}
 	}
 
-	if err := createKubeProxyConfigMap(cfg, localEndpoint, client); err != nil {
+	cmByte, err := createKubeProxyConfigMap(cfg, localEndpoint, client)
+	if err != nil {
 		return err
 	}
 
-	if err := createKubeProxyAddon(cfg, client); err != nil {
+	dsByte, err := createKubeProxyAddon(cfg, client)
+	if err != nil {
 		return err
 	}
 
-	if err := CreateRBACRules(client); err != nil {
-		return errors.Wrap(err, "error when creating kube-proxy RBAC rules")
+	if !PrintManifestFlag {
+		if err := CreateRBACRules(client); err != nil {
+			return errors.Wrap(err, "error when creating kube-proxy RBAC rules")
+		}
+	}
+
+	if PrintManifestFlag {
+		return printManifest(cmByte, dsByte, out)
 	}
 
 	fmt.Println("[addons] Applied essential addon: kube-proxy")
 	return nil
 }
 
+//  printManifest build manifests for installing kube-proxy and output the manifest to the out
+func printManifest(cmByte []byte, dsByte []byte, out io.Writer) error {
```

Comment:
i think we should add helper functions or variables that can be used for both:
- creating the objects with a client
- printing the objects

for example:
https://github.com/wangyysde/kubernetes/blob/edb2ff26a4cfbdead36bb55077b04727249d160d/cmd/kubeadm/app/phases/addons/proxy/proxy.go#L184-L189

this service account can be a variable in proxy.go called `serviceAccount`

similar all the roles and bindings from:
https://github.com/wangyysde/kubernetes/blob/edb2ff26a4cfbdead36bb55077b04727249d160d/cmd/kubeadm/app/phases/addons/proxy/proxy.go#L274-L331

can be variables.

then `printManifest()` can just marshalToYaml the objects from variables.

Reply:
if you look at the DNS printing we already have similar variables and we don't have to construct objects in two places.

Reply:
Ok, I will modify it.